sys/net/application_layer/dhcpv6/client.c

```diff
@@ -0,0 +1,932 @@
+    memcpy(sid->duid, server.duid.u8, server.duid_len);
+    return len + sizeof(dhcpv6_opt_t);
+}
+
+static inline uint16_t _get_elapsed_time(void)
+{
+    uint32_t now = _now_cs();
+    uint32_t elapsed_time = transaction_start - now;
+
+    if (elapsed_time > UINT16_MAX) {
+        /* xtimer_now_usec64() overflowed since transaction_start */
+        elapsed_time = (UINT32_MAX - transaction_start) + now + 1;
+    }
+    return elapsed_time;
+}
+
+static inline size_t _compose_elapsed_time_opt(dhcpv6_opt_elapsed_time_t *time)
+{
+    uint16_t len = 2U;
+
+    time->type = byteorder_htons(DHCPV6_OPT_ELAPSED_TIME);
+    time->len = byteorder_htons(len);
+    time->elapsed_time = byteorder_htons(_get_elapsed_time());
+    return len + sizeof(dhcpv6_opt_t);
+}
+
+static inline size_t _compose_oro_opt(dhcpv6_opt_oro_t *oro, uint16_t *opts,
+                                        unsigned opts_num)
+{
+    uint16_t len = 2U * opts_num;
+
+    oro->type = byteorder_htons(DHCPV6_OPT_ORO);
+    oro->len = byteorder_htons(len);
+    for (unsigned i = 0; i < opts_num; i++) {
+        oro->opt_codes[i] = byteorder_htons(opts[i]);
+    }
+    return len + sizeof(dhcpv6_opt_t);
+}
+
+static inline size_t _compose_ia_pd_opt(dhcpv6_opt_ia_pd_t *ia_pd,
+                                        uint32_t ia_id, uint16_t opts_len)
+{
+    uint16_t len = 12U + opts_len;
+
+    ia_pd->type = byteorder_htons(DHCPV6_OPT_IA_PD);
+    ia_pd->len = byteorder_htons(len);
+    ia_pd->ia_id = byteorder_htonl(ia_id);
+    ia_pd->t1.u32 = 0;
+    ia_pd->t2.u32 = 0;
+    return len + sizeof(dhcpv6_opt_t);
+}
+
+static inline size_t _add_ia_pd_from_config(uint8_t *buf)
+{
+    size_t msg_len = 0;
+
+    for (unsigned i = 0; i < DHCPV6_CLIENT_PFX_LEASE_MAX; i++) {
+        uint32_t ia_id = pfx_leases[i].parent.ia_id.id;
+        if (ia_id != 0) {
+            dhcpv6_opt_ia_pd_t *ia_pd = (dhcpv6_opt_ia_pd_t *)(&buf[msg_len]);
+
+            msg_len += _compose_ia_pd_opt(ia_pd, ia_id, 0U);
+        }
+    }
+    return msg_len;
+}
+
+static inline int32_t get_rand_us(void)
+{
+    int32_t rand = ((int32_t)random_uint32_range(0, 200 * US_PER_MS));
+    rand -= 100 * US_PER_SEC;
+    return rand;
+}
+
+static inline uint32_t _irt_us(uint16_t irt, bool greater_irt)
+{
+    uint32_t irt_us = (irt * US_PER_SEC);
+    int32_t rand = get_rand_us();
+
+    if (greater_irt && (rand < 0)) {
+        rand = -rand;
+    }
+    irt_us += (rand * irt_us) / US_PER_SEC;
+    return irt_us;
+}
+
+static inline uint32_t _sub_rt_us(uint32_t rt_prev_us, uint16_t mrt)
+{
+    uint32_t sub_rt_us = (2 * rt_prev_us) +
+                         ((get_rand_us() * rt_prev_us) / US_PER_SEC);
+
+    if (sub_rt_us > (mrt * US_PER_SEC)) {
+        uint32_t mrt_us = mrt * US_PER_SEC;
+
+        sub_rt_us = mrt_us + ((get_rand_us() * mrt_us) / US_PER_SEC);
+    }
+    return sub_rt_us;
+}
+
+#define OPT_LEN(opt)    (sizeof(dhcpv6_opt_t) + byteorder_ntohs((opt)->len))
+#define FOREACH_OPT(first, opt, len) \
```

Comment:
please make a function

Reply:
That is not possible... At least not in this quite easy-to-use and minimal way...

Reply:
not possible or you don't know how to?

Reply:
Well I could do something like

```C
static inline size_t iter_opt(dhcpv6_opt_t *prev, size_t msg_len, dhcpv6_opt_t **out)
{
     if (prev == NULL) {
         *out = prev;
     }
     else if (len > 0) {
         len -= OPT_LEN;
         *out = (dhcpv6_opt_t *)(((uint8_t *)prev + OPT_LEN(opt)))
     }
     return len;
}
```

But this is IMHO much uglier and possibly cost more ROM (there are pointer operations in here that did not show up above and I need to initialize `prev` somewhere) than the quite compact macro + I can't use it as a loop's head. Also, not to play the same string again, but this style was already [accepted by other maintainers](https://github.com/RIOT-OS/RIOT/pull/7014#discussion_r143329002). If you know of a nicer, equivalent transformation, let me know...

Reply:
Or simply factor out as much of the uglyness as possible, then use traditional loops. E.g.:

```
static size_t _opt_len(dhcpv6_opt_t *opt)
{
    return sizeof(dhcpv6_opt_t) + byteorder_ntohs(opt->len);
}

static dhcpv6_opt_t *_opt_next(dhcpv6_opt_t *opt)
{
    return (dhcpv6_opt_t *)((uint8_t *)opt + _opt_len(opt));
}

```

Then replace

```
FOREACH_OPT(first, opt, len) {
    /* loop body */
}
```

with sth like

```
for (opt = (void *)first; len; opt = _opt_next(opt)) {
    /* loop_body */
    len -= _opt_len(opt);
}
```

The point here is not whether something is "nicer", but what happens if

- someone looks at FOREACH_OPT(a,b,c) vs for (a, b, c) and has to understand what is happening

- if we use tools to scan / analyze / modify the codebase ...

Reply:
Fine. Done!